ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -260,4 +452,359 @@ private void updateMapHelper(Map<String, Long> map, String key, Long value) {
+        scanResults.set(newScanResults);
+      } catch (Exception e) {
+        logger.error("Exception while scanning index for stats bucketing", e);
+        metrics.blobStoreStatsErrorCount.inc();
+      } finally {
+        synchronized (scanLock) {
+          isScanning = false;
+          scanLock.notifyAll();
+        }
+        metrics.statsBucketingScanTotalTimeMs.update(time.milliseconds() - startTimeInMs);
+        metrics.statsBucketingScanCount.inc();
+      }
+    }
+
+    /**
+     * Perform a forward scan (order to newest entries) from the given start offset to the given end offset. This
+     * forward scan is used to process any new entries that were added to the index while scanning to the first
+     * checkpoint.
+     * @param startOffset the start offset defining the start of the scan range, inclusive (greater than or equal to)
+     * @param endOffset the end offset defining the end of the scan range, exclusive (strictly less than)
+     * @throws StoreException
+     */
+    private void forwardScan(Offset startOffset, Offset endOffset) throws StoreException {
+      SortedMap<Offset, IndexSegment> tailIndexSegments =
+          index.getIndexSegments().subMap(index.getIndexSegments().floorKey(startOffset), endOffset);
+      for (IndexSegment indexSegment : tailIndexSegments.values()) {
+        if (cancelled) {
+          return;
+        }
+        List<IndexEntry> indexEntries = getIndexEntries(indexSegment);
+        for (IndexEntry entry : indexEntries) {
+          IndexValue indexValue = entry.getValue();
+          if (indexValue.getOffset().compareTo(startOffset) < 0 || indexValue.getOffset().compareTo(endOffset) >= 0) {
+            // index value is either processed already or should not be processed
+            continue;
+          }
+          if (indexValue.isFlagSet(IndexValue.Flags.Delete_Index)) {
+            // delete record
+            IndexValue originalPut;
+            if (indexValue.getOriginalMessageOffset() != -1
+                && indexValue.getOriginalMessageOffset() >= indexSegment.getStartOffset().getOffset()) {
+              // delete and put are in the same index segment
+              originalPut = getDeletedPut(entry.getKey(), indexValue);
+              if (originalPut.getOffset().compareTo(startOffset) >= 0) {
+                processNewPut(newScanResults, originalPut);
+              }
+            } else {
+              Offset previousIndexSegmentOffset =
+                  index.getIndexSegments().size() > 1 ? index.getIndexSegments().lowerKey(indexSegment.getStartOffset())
+                      : index.getStartOffset();
+              FileSpan searchSpan = new FileSpan(index.getStartOffset(), previousIndexSegmentOffset);
+              originalPut = index.findKey(entry.getKey(), searchSpan);
+            }
+            if (originalPut != null && !originalPut.isFlagSet(IndexValue.Flags.Delete_Index)) {
+              // put exists and there are no multiple deletes
```

Comment:
so if there are multiple delete records, we might ignore the put record right. You don't search in a while loop until we find a put record or null. Not that I am suggesting. just wanted to confirm